### ISSUE TITLE
Support progress timeouts for SendStream

### ISSUE THREAD

**Issue by Nikratio** [python-trio/trio#414]
At the moment, the only way to send data is using `SendStream.send_all`. The problem is that this method blocks until everything has been sent. So if one wants to apply a timeout when no data is being sent at all (but not when data is merely being send very slowly), there is no way to do that. It would be nice if `SendStream` could either get a `send_some` method (that returns as soon as some data has been sent), or if the `SendStream.send_all` method could gain a `timeout` parameter.

**Comment by njsmith** [comment]
That's an interesting point that I've been thinking about all day :-).

I *think* in practice this isn't as useful as it sounds.

So obviously you can do something like, say "I want to make sure that they read at a minimum of X bytes/second", and then use that to set a timeout on `send_all`. (You can get fancier, like giving some grace period for temporary connection hiccups or aggregating over multiple calls, but that's the basic idea.)

You want to distinguish between "no data sent at all" and "data being sent slowly". And intuitively, this makes sense given how `send`'s API works: sometimes it returns a partial success, and sometimes it blocks entirely. And certainly putting timeouts on `send` calls is a pragmatic way to make sure that if the other side does stop reading entirely, it will eventually time out. But if you look at the details of how `send` actually works, the distinction's not that simple. What `send` actually does is hand off data to the kernel's send buffer, and the kernel has some rules for when it accepts a `send` or not.

Windows as usual is weird: IIRC its send buffer is basically "two calls to `send`". If the data from your last `send` is still trickling out, then this call to `send` will block until that's finished, and then it will succeed all at once. AFAICT Windows never does partial `send`s. So in this case, putting a timeout on `send_all` is equivalent to putting a timeout on individual `send` calls.

On the Unixes, there's a buffer that has some size, which in modern implementations is generally adjusted on the fly by some auto-tuning algorithm (you want bigger buffers for connections that have high bandwidth and/or latency). And then there's some rule for when it will allow `send` to succeed. If I'm reading the kernel source right, it looks like a socket will accept data up to the size of its send buffer, but once the buffer hits capacity then it stops accepting new data until it reaches half its maximum capacity. So if the buffer is 2 megabytes, then eventually it fills up, and the next call to `send` after that will block until the other side has read and acknowledged 1 MB of data. If it's reading at, say, 1 KB/s, this will take 1000 seconds, so the timeout on your `send` call might well fire – even though the other side is reading continuously the whole time as fast as it can. So the point is that there's actually no way to distinguish between "no data being sent at all" versus "data ... merely being sen[t] very slowly" – by putting a timeout on individual `send` calls, you're actually imposing some kind of minimum requirement on how fast data has to be sent. But you don't actually know what this requirement is, because it depends on the size of the kernel's send buffer, which changes all the time.

(On recent Linux and MacOS, trio also generally enables TCP_NOTSENT_LOWAT, which complicates these analyses even more – it makes the requirement for `send` succeeding dependent on both how much data has been sent, and how much data has been acknowledged.)

It is true that putting a fixed timeout on `send_all` might work a bit worse than putting a fixed timeout on `send` (at least on Unix), because if the buffer you're trying to send it large enough you may end up waiting for the first `send` to succeed + some extra `send`s for the rest of the buffer. So the optimal timeout becomes dependent on not just how much data you've sent before (like for regular `send`), but also how much data you're sending now – so if you're putting a timeout on `send_all` instead of `send`, maybe it becomes more important to scale the timeout to be bigger for larger buffers. This is really the easiest part of this problem, though; if you're OK with your effective minimum-transmit-speed being randomly determined by the whims of the kernel and the detailed history of your previous calls to `send`, then you can probably cope with this extra bit of uncertainty or adjusting your timeouts.

Implementing `send_some` is also basically impossible for `SSLStream`, because you have to encrypt the data and then call `send` on the underlying socket, but if `send` only accepts part of the data, then you're stuck: you can't go back and "unencrypt" the rest of the data, you've already updated the SSL/TLS state machine and are stuck with it. So in practice a `SSLStream.send_some` would have to just be implemented as `send_all`, which means that any protocol implementation that wants to work over encrypted streams would need to have its timeout handling written in a way that works for `send_all` *anyway*, and that's basically every protocol these days.

Also, for what it's worth, checking the implementation of CPython's `socket.socket.sendall` method, it looks like there the socket timeout is applied to the whole `sendall` call, not to the individual `send` operations inside it. And this is what popular libraries like `requests` ultimately use to implement their timeouts. So you may have already been using this strategy without realizing it.

So I don't think there are major advantages to implementing a `send_some` method or adding a `timeout` parameter to `send_all`. What I *would* like to see is some prepackaged way to set a minimum-transmit-rate, so you could say "okay, my peer has to receive at 1 KB/s", and it would automatically calculate timeouts on individual `send_all` calls – maybe as a `Stream` wrapper. Doing it right is tricky because of the issues described above, but nonetheless it's a really common problem and there should be a common solution, even if it isn't perfect. And I'm sure we could do better than the simple naive thing.

This blog post from CloudFlare is really interesting on these issues (and describes a case they ran into in production where the standard "put a timeout on `send`" strategy goes wrong because of the issues above): https://blog.cloudflare.com/the-curious-case-of-slow-downloads/

**Comment by Nikratio** [comment]
Wow, that's interesting stuff! Thanks for writing it all up.

According to the Cloudflare post, at least one way to achieve what I want is to combine a timeout which checking if the kernel's send buffer is actually getting drained via `ioctl`. 

I'm not sure if I'd really expect this kind of functionality in Trio, but I am sure that I wish that Trio would allow me to easily terminate stale connections without cutting of slow connections :-P.

**Comment by njsmith** [comment]
> According to the Cloudflare post, at least one way to achieve what I want is to combine a timeout which checking if the kernel's send buffer is actually getting drained via `ioctl`.

Yeah, it's tricky though, because that's a Linux-only API, and it only works for raw sockets. (And maybe only TCP sockets, I'm not sure.) Could be useful, and you could implement a version of `SocketStream` that does that using Trio, but it's not obviously The One And Only Right Way To Do It.

> I am sure that I wish that Trio would allow me to easily terminate stale connections without cutting of slow connections :-P.

The tl;dr of my post above is that what you're asking for here isn't really a thing that exists, but you can get the same behavior that every other Python library uses by putting a timeout on `send_all`.

**Comment by njsmith** [comment]
Re: the portability of checking the send buffer, apparently [I researched this before](https://stackoverflow.com/a/40207176/) :-). (I think while trying to understand TCP_NOTSENT_LOWAT.)
